diff --git a/lld/ELF/Driver.cpp b/lld/ELF/Driver.cpp
index c2059c70e..181e4241b 100644
--- a/lld/ELF/Driver.cpp
+++ b/lld/ELF/Driver.cpp
@@ -1832,6 +1832,13 @@ void LinkerDriver::createFiles(opt::InputArgList &args) {
     }
   }
 
+  // Sort files by filename.
+  // This makes the linker produce the same binary even when the order of
+  // command line arguments changes.
+  std::sort(files.begin(), files.end(), [](InputFile *lhs, InputFile *rhs) {
+    return lhs->getName() < rhs->getName();
+  });
+
   if (files.empty() && !hasInput && errorCount() == 0)
     error("no input files");
 }
